pick.c

```diff
@@ -1041,13 +1063,55 @@ get_key(const char **key)
 	return PRINTABLE;
 }
 
+enum key
+get_vi_key(const char **key)
+{
+	static struct {
+		const char	*str;
+		enum key	 key;
+		int		 vi_mode;
+	}			keys[] = {
+		{ " ",		RIGHT,		1 },
+		{ "$",		CTRL_E,		1 },
+		{ "/",		UNKNOWN,	0 },
+		{ "0",		CTRL_A,		1 },
+		{ "A",		CTRL_E,		0 },
+		{ "a",		RIGHT,		0 },
+		{ "D",		CTRL_K,		1 },
+		{ "d",		CTRL_U,		1 },
+		{ "G",		END,		1 },
+		{ "H",		HOME,		1 },
+		{ "h",		LEFT,		1 },
+		{ "I",		CTRL_A,		0 },
+		{ "i",		UNKNOWN,	0 },
+		{ "j",		LINE_DOWN,	1 },
+		{ "k",		LINE_UP,	1 },
+		{ "l",		RIGHT,		1 },
+		{ "Q",		CTRL_C,		0 },
+		{ "x",		DEL,		1 },
+		{ "Z",		ALT_ENTER,	0 },
+		{ NULL,		UNKNOWN,	0 },
+	};
+	int			 i;
+
+	for (i = 0; keys[i].str != NULL; i++) {
+		if (strcmp(keys[i].str, *key) == 0) {
+			vi_mode = keys[i].vi_mode;
+			return keys[i].key;
+		}
+	}
+
+	return UNKNOWN;
+}
+
 int
 tty_getc(void)
 {
-	int	c;
+	int	c, n;
 
-	if ((c = getc(tty_in)) == ERR && !gotsigwinch)
-		err(1, "getc");
+	while ((n = read(fileno(tty_in), &c, 1)) == 0);
+	if (n == ERR && !gotsigwinch)
```

Comment:
I would prefer `-1` instead of `ERR` for consistency with the rest of the code. Also, continuing reading on `errno == EINTR` would be a good addition.

Reply:
Changed it to:
```C
again:
        while ((n = read(fileno(tty_in), &c, 1)) == 0);
        if (n == -1) {
                if (errno == EINTR)
                        goto again;
                if (!gotsigwinch)
                        err(1, "read");
        }

        return c;
```
OK?

Reply:
I would rather loop until success/error, untested code below:

I would rather loop until success/error, untested code below:

```c
for (;;) {
  n = read(fileno(tty_in), &c, 1);
  if (n == -1) {
    if (errno == EINTR)
      continue;
    if (gotsigwinch)
      break;
    err(1, "read");
  }
  if (n > 0)
    break;
}
```